pep-0xxx.rst

```diff
@@ -0,0 +1,571 @@
+            """
+            pass
+
+        @abstractmethod
+        def set_inner_protocols(self, protocols: List[str]) -> None:
+            """
+            Specify which protocols the socket should advertise as supported
+            during the TLS handshake. This may be advertised using either or
+            both of ALPN or NPN.
+
+            ``protocols`` should be a list of ASCII strings for ALPN tokens,
+            such as ``['h2', 'http/1.1']``, ordered by preference. The
+            selection of the protocol will happen during the handshake, and
+            will use whatever protocol negotiation mechanisms are available and
+            supported by both peers.
+
+            If the TLS implementation doesn't support protocol negotiation,
+            this method will raise ``NotImplementedError``.
+            """
+            pass
+
+        @abstractmethod
+        def set_sni_callback(self, callback: Optional[ServerNameCallback]) -> None:
+            """
+            Register a callback function that will be called after the TLS
+            Client Hello handshake message has been received by the TLS server
+            when the TLS client specifies a server name indication.
+
+            Only one callback can be set per ``Context``. If ``callback`` is
+            ``None`` then the callback is disabled. Calling this function a
+            subsequent time will disable the previously registered callback.
+
+            The ``callback`` function will be called with three arguments: the
+            first will be the ``TLSBufferObject`` for the connection; the
+            second will be a string that represents the server name that the
+            client is intending to communicate (or ``None`` if the TLS Client
+            Hello does not contain a server name); and the third argument will
+            be the original ``Context``. The server name argument will be the
+            IDNA *decoded* server name.
+
+            The ``callback`` must return ``None`` to allow negotiation to
+            continue. Other return values signal errors. Attempting to control
+            what error is signaled by the underlying TLS implementation is not
+            specified in this API, but is up to the concrete implementation to
+            handle.
+            """
+            pass
+
+        @abstractmethod
+        def wrap_socket(self, socket: socket.socket, server_side=False: bool,
+                        auto_handshake=True: bool,
+                        server_hostname=None: Optional[str]) -> TLSWrappedSocket:
+            """
+            Wrap an existing Python socket object ``socket`` and return a
+            ``TLSWrappedSocket`` object. ``socket`` must be a ``SOCK_STREAM``
+            socket: all other socket types are unsupported.
+
+            The returned SSL socket is tied to the context, its settings and
+            certificates.
+
+            The parameter ``server_side`` is a boolean which identifies whether
```

Comment:
FWIW this looks like a kludge that could be prevented by having separate Context classes for clients and servers.  The more I think about it the more I like it.